web-app/test/unit/resources/elements/panels/main-panel-spec.js

```diff
@@ -114,34 +114,80 @@ describe('MainPanel', () => {
 
     describe('setScalingFactor', () => {
 
+        const dataUri = 'data:image/jpeg;base64,R0lGODlhkAGQAfcAAAQCBJKDLJSSlExGHNS/NNTOtFRSVCckFGxjJPjgXMzKzHyCfKuqpDMxNPzu0fXaJGxqZLikLEdCPBQSDDQyFOvq7H90LL6PlFxOw==';
+
+        const mockImage = {
+            src: null,
+            onload: () => {}
+        };
+
+        const realImage = global.Image;
+
+        beforeAll(() => {
+            global.Image = () => { return mockImage; };
+        });
+
+        afterAll(() => {
+            global.Image = realImage;
+        })
+
         beforeEach(() => {
-            // need to mock the config setting: this.options.image.fitImageToWindow
-            mainpanel.offsetWidth = 100;
-            mainpanel.offsetHeight = 100;
+            mainpanel.scalingFactor = 2.0; // dummy factor
+            mainpanel.options = {};
+            mainpanel.dataURI = dataUri;
         });
 
-        it('verify minimum zoom of 0.125', async () => {
-            // mock image dimension
+        it('verify no scaling calculation without option set', () => {
             mainpanel.setScalingFactor();
-            expect(mainpanel.scalingFactor).toBe(0.125);
+            mockImage.onload();
+            expect(mainpanel.scalingFactor).toBeCloseTo(2.0, 2);
         });
 
-        it('verify maximum zoom of 1', async () => {
-            // mock image dimension
+        it('scaling calculation run on image load with option set', () => {
+            mainpanel.imageCanvas = {offsetWidth: 400, offsetHeight: 400}; // current onscreen canvas size
+            _.set(mainpanel.options, 'image.fitImageToWindow', true);
+            mockImage.width = 700; // less than twice the size of the canvas
+            mockImage.height = 200;
             mainpanel.setScalingFactor();
-            expect(mainpanel.scalingFactor).toBe(1);
+            mockImage.onload(); // we need to ensure onload is called since src is not really triggering it
+
+            expect(mainpanel.scalingFactor).toBeCloseTo(0.5, 2);
         });
 
-        it('verify zoom to fit wide image', async () => {
-            // mock image dimension
-            mainpanel.setScalingFactor();
-            expect(mainpanel.scalingFactor).toBe(0.25);
+    });
+
+    describe('getScalingFactorFromImage', () => {
+
+        let elem = {};
+
+        beforeEach(() => {
+            elem.offsetWidth = 1024;
+            elem.offsetHeight = 1024;
         });
 
-        it('verify zoom to fit tall image', async () => {
-            // mock image dimension
-            mainpanel.setScalingFactor();
-            expect(mainpanel.scalingFactor).toBe(0.5);
+        it('verify minimum zoom of 0.125', () => {
+            // more than 8x the width
+            let image = {height: 2048, width: 8600};
+            let factor = mainpanel.getScalingFactorFromImage(elem, image);
```

Comment:
with floating point you can't do 1.0 === 1.0 because it might actually be 1.00000001 so we use the jest expect function toBeCloseTo which will determine to a certain number of decimal places (in this case 4.  So 0.125, 0.1257 would all equate to 0.125.  Of course if the result was 0.1257 we have bigger problems of bad CPU arithmetic.